In [ ]:
def x_1(d_1, d_2):
    return 100 + .5*(d_1 - 430) - .25*(d_2 - 460)

def x_3(d_2):
    return 230 + .5*(d_2 - 460)

def x_6(d_1, d_2, d_3):
    return 20 - 2*(d_1 - 430) + (d_2 - 460) + (d_3 - 420)

In [ ]:
d_1 = 90
d_2 = 4
d_3 = 100
print(x_1(d_1=d_1, d_2=d_2), x_3(d_2=d_2), x_6(d_1=d_1, d_2=d_2, d_3=d_3))

44.0 2.0 -76


In [ ]:
d_1 = 90
d_2 = 4
d_3 = 100


x1 = x_1(d_1=d_1, d_2=d_2)
x3 = x_3(d_2=d_2)
x6 = x_6(d_1=d_1, d_2=d_2, d_3=d_3)


if x1 < 0:
    print("Es un valor factible")
else:
    print("Valor no factible")
if x3 < 0:
    print("Es un valor factible")
else:
    print("Valor no factible")
if x6 < 0:
    print("Es un valor factible")
else:
    print("Valor no factible")

Valor no factible
Valor no factible
Es un valor factible


In [ ]:
valores_invalidos = []
valores_validos = []
for i_1 in range(1, 100):
    for i_2 in range(1, 100):
        for i_3 in range(1, 100):
            d_1 = i_1
            d_2 = i_2
            d_3 = i_3


            x1 = x_1(d_1=d_1, d_2=d_2)
            x3 = x_3(d_2=d_2)
            x6 = x_6(d_1=d_1, d_2=d_2, d_3=d_3)


            if x1 > 0:
                pass
            else:
                valores_invalidos.append((d_1, d_2, d_3))

            if x3 > 0:
                pass
            else:
                valores_invalidos.append((d_1, d_2, d_3))

            if x6 > 0:
                pass
            else:
                valores_invalidos.append((d_1, d_2, d_3))
            if x1 > 0 and x3 > 0 and x6 >0:
                valores_validos.append((d_1, d_2, d_3))
print("d_1, d_2, d_3")
print(d_1, d_2, d_3)

d_1, d_2, d_3
99 99 99


In [ ]:
valores_validos

In [ ]:
len(valores_validos)

240149

In [ ]:
d_1 = 4
d_2 = 2
d_3 = 41


x1 = x_1(d_1=d_1, d_2=d_2)
x3 = x_3(d_2=d_2)
x6 = x_6(d_1=d_1, d_2=d_2, d_3=d_3)



Valor no factible
Valor no factible
Valor no factible


In [ ]:
d_1 = 4
d_2 = 2
d_3 = 41
x1 = x_1(d_1=d_1, d_2=d_2)
x3 = x_3(d_2=d_2)
x6 = x_6(d_1=d_1, d_2=d_2, d_3=d_3)
print("x1, x3, x6")
print(x1, x3, x6)

x1, x3, x6
1.5 1.0 35


Optimizar el siguiente problema:


$$\text{Max  } Z = 450x_1 + 120x_2 + 22x_3 +28x_4 - 28x_5$$

$$
0.1x_1 + x_2  +x_4 \leq 800
$$

$$
0.05x_1 + 2x_2 + 3x_4  \leq 1000
$$

$$
5x_1 - x_3 + 4x_4 + x_5  = 0
$$


Donde $x_4$ son los costos de la empresa.

In [ ]:
! pip install cplex
! pip install docplex


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.0/57.0 kB 1.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.3/44.3 MB 9.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 645.5/645.5 kB 18.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for docplex: filename=docplex-2.29.241-py3-none-any.whl size=685423 sha256=0b1663395d9ece5d94d52351460ebd2af6f80b56c483419ee62f6f0a4b4bf705
  Stored in directory: /root/.cache/pip/wheels/00/4f/37/fbc42df1f9b6786024d4b39c81f0b373c84e6fab9e18513fc2
Successfully built docplex


In [ ]:
import os
# os.environ['CPLEX_STUDIO_BINARIES'] = '/path/to/cplex/bin'


In [ ]:
# Cargar CPLEX
from docplex.mp.model import Model

In [ ]:
# Iniciar el modelo
model = Model(name='My_model')

In [ ]:
# Crear variables.
x_1 = model.continuous_var(name='x_1')
x_2 = model.continuous_var(name='x_2')
x_3 = model.continuous_var(name='x_3')
x_4 = model.continuous_var(name='x_4')
x_5 = model.continuous_var(name='x_5')




$$\text{Max  } Z = 450x_1 + 120x_2 + 22x_3 +28x_4 - 28x_5$$

$$
0.1x_1 + x_2  +x_4 \leq 800
$$

$$
0.05x_1 + 2x_2 + 3x_4  \leq 1000
$$

$$
5x_1 - x_3 + 4x_4 + x_5  = 0
$$


In [ ]:
# Definir la función objetivo:
model.maximize(450*x_1 + 120*x_2 + 22*x_3 + 28*x_4 - 28*x_5)
# Definir las restricciones:
model.add_constraint(0.1*x_1 + x_2 + x_4 <= 800)
model.add_constraint(0.05*x_1 + 2*x_2 + 3*x_4 <= 1000)
model.add_constraint(5*x_1 - x_3 + 4*x_4 + x_5 == 0)
# Solucionar el modelo :
solution = model.solve(log_output=1)

In [ ]:
solution.display()

solution for: My_model
objective: 4480000.000
status: OPTIMAL_SOLUTION(2)
x_1 = 8000.000
x_3 = 40000.000


In [ ]:
list_contrains = []
for i in range(model.number_of_constraints):
    c = model.get_constraint_by_index(i)
    list_contrains.append(c)
    print(c)

0.100x_1+x_2+x_4 <= 800
0.050x_1+2x_2+3x_4 <= 1000
5x_1-x_3+4x_4+x_5 == 0


In [ ]:
# Imprimamos información sobre las variables de holgura:
for i, h in enumerate(model.slack_values(list_contrains)):
    print(f"Holgura para la restricción {i+1}: {h}")




Holgura para la restricción 1: 0
Holgura para la restricción 2: 600.0
Holgura para la restricción 3: 0


In [ ]:
# Imprimamos información sobre los precios sobra o variables duales:
for i, d in enumerate(model.dual_values(list_contrains)):
    print(f"Precio sombra para la restricción {i+1}: {d}")


Precio sombra para la restricción 1: 5600.0
Precio sombra para la restricción 2: 0
Precio sombra para la restricción 3: -22.0


In [ ]:
# Obrener iinformación de la sensibilidad:

cpx = model.get_engine().get_cplex()
sensitivity_o_f = cpx.solution.sensitivity.objective()
print("Intervalos donde puedo cambiar los parámetros de las x_s, sin que afecten el valor de la función objetivo")
for i, x_i in enumerate(sensitivity_o_f):
    print(f"Intervalo de x_{i+1}: [{x_i[0]}, {x_i[1]}]")



Intervalos donde puedo cambiar los parámetros de las x_s, sin que afecten el valor de la función objetivo
Intervalo de x_1: [-98.0, 1e+20]
Intervalo de x_2: [-1e+20, 5600.0]
Intervalo de x_3: [-87.6, 28.0]
Intervalo de x_4: [-1e+20, 5512.0]
Intervalo de x_5: [-1e+20, -22.0]


In [ ]:
b_contrains = cpx.solution.sensitivity.rhs()
print("Precio sombra (dual) y su rango donde es valido.")
for i, (d, rhs) in enumerate(zip(model.dual_values(list_contrains), b_contrains)):
    print(f"Restricción {i+1}, precio sombra: {d}, intervalo: [{rhs[0]}, {rhs[1]}]")



Precio sombra (dual) y su rango donde es valido.
Restricción 1, precio sombra: 5600.0, intervalo: [0.0, 2000.0]
Restricción 2, precio sombra: 0, intervalo: [400.0, 1e+20]
Restricción 3, precio sombra: -22.0, intervalo: [-1e+20, 40000.0]


Actividad en clase:
Realiza el análisis de sensibilidad para el siguiente caso:
### Función objetivo

$$
\text{Max } Z = 3x_1 + 4x_2
$$
Restricciones de actividades:
$$
2x_1 + x_2 \leq 100
$$
$$
x_1 + 3x_2 \leq 90
$$
Restricciones de no negatividad:

$$
x_1 \geq 0
$$
$$
x_2 \geq 0
$$

In [ ]:
import os
from docplex.mp.model import Model

model = Model(name='Problema_Maximizacion')


x_1 = model.continuous_var(name='x_1', lb=0)
x_2 = model.continuous_var(name='x_2', lb=0)

# función objetivo
model.maximize(3*x_1 + 4*x_2)

# Definir las restricciones
c1 = model.add_constraint(2*x_1 + x_2 <= 100, 'restriccion_1')
c2 = model.add_constraint(x_1 + 3*x_2 <= 90, 'restriccion_2')

# Lista de restricciones
list_contrains = [c1, c2]

# Solucionar el modelo
solution = model.solve(log_output=True)

# Mostrar la solución
if solution:
    print("\nSolución encontrada:")
    print(f"Valor óptimo de Z = {solution.objective_value}")
    print(f"x_1 = {solution.get_value(x_1)}")
    print(f"x_2 = {solution.get_value(x_2)}")

    # Mostrar información sobre las restricciones
    print("\nInformación de las restricciones:")
    for i, c in enumerate(list_contrains):
        print(c)

    # Información sobre las variables de holgura
    print("\nVariables de holgura:")
    for i, h in enumerate(model.slack_values(list_contrains)):
        print(f"Holgura para la restricción {i+1}: {h}")

    # Información sobre los precios sombra (variables duales)
    print("\nPrecios sombra:")
    for i, d in enumerate(model.dual_values(list_contrains)):
        print(f"Precio sombra para la restricción {i+1}: {d}")

    # Análisis de sensibilidad
    cpx = model.get_engine().get_cplex()

    # Sensibilidad de los coeficientes de la función objetivo
    sensitivity_o_f = cpx.solution.sensitivity.objective()
    print("\nIntervalos donde puedo cambiar los parámetros de las x_s sin que afecten la solución óptima:")
    for i, x_i in enumerate(sensitivity_o_f):
        print(f"Coeficiente de x_{i+1}: [{x_i[0]}, {x_i[1]}]")

    # Sensibilidad de los términos independientes de las restricciones
    b_contrains = cpx.solution.sensitivity.rhs()
    print("\nPrecios sombra (dual) y su rango donde es válido:")
    for i, (d, rhs) in enumerate(zip(model.dual_values(list_contrains), b_contrains)):
        print(f"Restricción {i+1}, precio sombra: {d}, intervalo: [{rhs[0]}, {rhs[1]}]")
else:
    print("No se encontró solución.")

Version identifier: 22.1.2.0 | 2024-12-10 | f4cec290b
CPXPARAM_Read_DataCheck                          1
Tried aggregator 1 time.
No LP presolve or aggregator reductions.
Presolve time = 0.00 sec. (0.00 ticks)

Iteration log . . .
Iteration:     1   Dual infeasibility =             0.000000
Iteration:     2   Dual objective     =           190.000000

Solución encontrada:
Valor óptimo de Z = 190.0
x_1 = 42.0
x_2 = 16.0

Información de las restricciones:
restriccion_1: 2x_1+x_2 <= 100
restriccion_2: x_1+3x_2 <= 90

Variables de holgura:
Holgura para la restricción 1: 0
Holgura para la restricción 2: 0

Precios sombra:
Precio sombra para la restricción 1: 1.0
Precio sombra para la restricción 2: 1.0

Intervalos donde puedo cambiar los parámetros de las x_s sin que afecten la solución óptima:
Coeficiente de x_1: [1.3333333333333333, 8.0]
Coeficiente de x_2: [1.5, 9.0]

Precios sombra (dual) y su rango donde es válido:
Restricción 1, precio sombra: 1.0, intervalo: [30.0, 180.0]
Restricción

In [ ]:
# Iniciar el modelo
model = Model(name='My_model')

# Crear variables.
x_1 = model.continuous_var(name='x_1')
x_2 = model.continuous_var(name='x_2')


# Definir la función objetivo:
model.maximize(3*x_1 + 4*x_2 )
# Definir las restricciones:
model.add_constraint(2*x_1 + x_2  <= 100)
model.add_constraint(x_1 + 3*x_2 <= 90)
# Solucionar el modelo :
solution = model.solve(log_output=1)

solution.display()

list_contrains = []
for i in range(model.number_of_constraints):
    c = model.get_constraint_by_index(i)
    list_contrains.append(c)
    print(c)

# Imprimamos información sobre las variables de holgura:
for i, h in enumerate(model.slack_values(list_contrains)):
    print(f"Holgura para la restricción {i+1}: {h}")

# Imprimamos información sobre los precios sobra o variables duales:
for i, d in enumerate(model.dual_values(list_contrains)):
    print(f"Precio sombra para la restricción {i+1}: {d}")


# Obrener iinformación de la sensibilidad:

cpx = model.get_engine().get_cplex()
sensitivity_o_f = cpx.solution.sensitivity.objective()
print("Intervalos donde puedo cambiar los parámetros de las x_s, sin que afecten el valor de la función objetivo")
for i, x_i in enumerate(sensitivity_o_f):
    print(f"Intervalo de x_{i+1}: [{x_i[0]}, {x_i[1]}]")

b_contrains = cpx.solution.sensitivity.rhs()
print("Precio sombra (dual) y su rango donde es valido.")
for i, (d, rhs) in enumerate(zip(model.dual_values(list_contrains), b_contrains)):
    print(f"Restricción {i+1}, precio sombra: {d}, intervalo: [{rhs[0]}, {rhs[1]}]")






solution for: My_model
objective: 190.000
status: OPTIMAL_SOLUTION(2)
x_1 = 42.000
x_2 = 16.000
2x_1+x_2 <= 100
x_1+3x_2 <= 90
Holgura para la restricción 1: 0
Holgura para la restricción 2: 0
Precio sombra para la restricción 1: 1.0
Precio sombra para la restricción 2: 1.0
Intervalos donde puedo cambiar los parámetros de las x_s, sin que afecten el valor de la función objetivo
Intervalo de x_1: [1.3333333333333333, 8.0]
Intervalo de x_2: [1.5, 9.0]
Precio sombra (dual) y su rango donde es valido.
Restricción 1, precio sombra: 1.0, intervalo: [30.0, 180.0]
Restricción 2, precio sombra: 1.0, intervalo: [50.0, 300.0]


Holgura para la restricción 1: 0: La primera restricción se cumple con igualdad
Holgura para la restricción 2: 0: La segunda restricción también se cumple con igualdad

Precios sombra:

Precio sombra para la restricción 1: 1.0: Si aumentamos el límite de la primera restricción (100) en 1 unidad, el valor óptimo de Z aumentará en 1 unidad.
Precio sombra para la restricción 2: 1.0: Si aumentamos el límite de la segunda restricción (90) en 1 unidad, el valor óptimo de Z aumentará en 1 unidad.

Análisis de sensibilidad para los coeficientes de la función objetivo:

Intervalo de x_1: [1.3333333333333333, 8.0]: El coeficiente de x₁ (actualmente 3) puede variar entre 1.33 y 8 sin que cambie la solución óptima actual (los valores de x₁ y x₂).
Intervalo de x_2: [1.5, 9.0]: El coeficiente de x₂ (actualmente 4) puede variar entre 1.5 y 9 sin que cambie la solución óptima actual.

Análisis de sensibilidad para los términos independientes:

Restricción 1, precio sombra: 1.0, intervalo: [30.0, 180.0]: El precio sombra de 1.0 es válido mientras el lado derecho de la primera restricción (100) varíe entre 30 y 180.
Restricción 2, precio sombra: 1.0, intervalo: [50.0, 300.0]: El precio sombra de 1.0 es válido mientras el lado derecho de la segunda restricción (90) varíe entre 50 y 300